subsys/net/lib/nrf_cloud/src/nrf_cloud_logs.c

```diff
@@ -0,0 +1,452 @@
+			return -EBUSY;
+		}
+	}
+
+	/* For REST, assume is ready. */
+	return 0; /* Logging is ready. */
+}
+
+static int logger_format_set(const struct log_backend *const backend, uint32_t log_type)
+{
+	if (backend != &log_nrf_cloud_backend) {
+		return 0;
+	}
+	if ((log_type == LOG_OUTPUT_TEXT) || (log_type == LOG_OUTPUT_DICT)) {
+		log_format_current = log_type;
+		return 0;
+	}
+	return -ENOTSUP;
+}
+
+static void logger_notify(const struct log_backend *const backend, enum log_backend_evt event,
+		       union log_backend_evt_arg *arg)
+{
+	if (backend != &log_nrf_cloud_backend) {
+		return;
+	}
+
+	if (event == LOG_BACKEND_EVT_PROCESS_THREAD_DONE) {
+		/* Flush our transmission buffer */
+		send_ring_buffer();
+
+		LOG_INF("Logged lines:%u, bytes:%u; buf bytes:%u; sent lines:%u, sent bytes:%u",
+			stats.lines_rendered, stats.bytes_rendered,
+			ring_buf_size_get(&log_nrf_cloud_rb),
+			stats.lines_sent, stats.bytes_sent);
+	}
+}
+
+static int send_ring_buffer(void)
+{
+	int err;
+	int ret;
+	struct nrf_cloud_tx_data output = {
+		.qos = MQTT_QOS_0_AT_MOST_ONCE,
+		.topic_type = (log_format_current == LOG_OUTPUT_TEXT) ?
+			       NRF_CLOUD_TOPIC_BULK : NRF_CLOUD_TOPIC_BIN
+	};
+	uint32_t stored;
+	uint8_t *base64_buf = NULL;
+
+	/* The bulk topic requires the multiple JSON messages to be placed in
+	 * a JSON array. Close the array then send it.
+	 */
+	if ((num_msgs != 0) && (log_format_current == LOG_OUTPUT_TEXT)) {
+		ring_buf_put(&log_nrf_cloud_rb, "]", 1);
+	}
+
+	stored = ring_buf_size_get(&log_nrf_cloud_rb);
+	output.data.len = ring_buf_get_claim(&log_nrf_cloud_rb,
+					     (uint8_t **)&output.data.ptr,
+					     stored);
+	if (output.data.len != stored) {
+		LOG_WRN("Capacity:%u, free:%u, stored:%u, claimed:%u",
+			ring_buf_capacity_get(&log_nrf_cloud_rb),
+			ring_buf_space_get(&log_nrf_cloud_rb),
+			stored, output.data.len);
+		stored = output.data.len;
+	}
+
+	if ((log_format_current == LOG_OUTPUT_DICT) &&
+	    IS_ENABLED(CONFIG_NRF_CLOUD_LOG_DICT_JSON)) {
+		/* Convert to base64 and wrap in JSON */
+		size_t olen;
+		size_t buflen;
+
+		/* Determine buffer size needed */
+		base64_encode(NULL, 0, &olen, output.data.ptr, output.data.len);
+		buflen = strlen(JSON_FMT1) + olen + strlen(JSON_FMT2);
+
+		base64_buf = k_calloc(buflen, 1);
+		if (base64_buf == NULL) {
+			err = -ENOMEM;
+			goto cleanup;
+		}
+
+		memcpy(base64_buf, JSON_FMT1, strlen(JSON_FMT1));
+		err = base64_encode(&base64_buf[strlen(JSON_FMT1)], olen, &olen,
+				    output.data.ptr, output.data.len);
+		if (err) {
+			goto cleanup;
+		}
+		memcpy(&base64_buf[strlen(JSON_FMT1) + olen], JSON_FMT2, strlen(JSON_FMT2));
+		output.data.ptr = base64_buf;
+		output.data.len = buflen - 1;
+	} /* Else send ring buffer contents as is -- JSON or raw binary. */
+
+	if (IS_ENABLED(CONFIG_NRF_CLOUD_MQTT)) {
+		err = nrf_cloud_send(&output);
+	} else if (IS_ENABLED(CONFIG_NRF_CLOUD_REST)) {
+		/* send to d2c topic, unless bulk is true, in which case d2c/bulk is used */
```

Comment:
Since I hadn't concentrated on REST support, I missed that when I added bulk support for MQTT.  This should use bulk.  I also need to add bin support to the REST call.

Reply:
Using bulk on REST is working now.